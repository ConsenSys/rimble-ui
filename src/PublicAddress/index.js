import React, { Component } from 'react';
import styled, { keyframes } from 'styled-components';

import Box from '../Box';
import Button from '../Button';
import Input from '../Input';
import Field from '../Field';

const slideIn = keyframes`
  from {
    opacity: 0;
    transform: translateY(100%);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
`;

const fadeOut = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

const StyledInput = styled(Input.InputOnly)`
  text-overflow: ellipsis;
  white-space: no-wrap;
`;

const StyledButton = styled(Button)`
  & {
    position: relative;
    overflow: hidden;

    display: flex;
    align-items: center;
    justify-content: center;
  }

  &.-is-copied::before {
    content: '';
    position: absolute;

    width: 100%;
    height: 100%;
    background: inherit;
    opacity: 1;

    animation-name: ${fadeOut};
    animation-duration: 100ms;
    animation-timing-function: linear;
  }

  &::after {
    opacity: 0;
    content: 'Copied!';
    position: absolute;
    width: 100%;
    background: inherit;
    transition: all 500ms ease;
  }

  &.-is-copied::after {
    opacity: 1;
    animation-name: ${slideIn};
    animation-duration: 1000ms;
    animation-timing-function: ease;
  }
`;

const StyledWrapper = styled(Box)`
  & {
    display: flex;
    flex-direction: row;
    align-items: center;
    width: 100%;
    position: relative;
  }

  > input {
    width: 100%;
    padding-right: 5.25rem;
  }

  > button {
    position: absolute;
    right: 0.75rem;
  }
`;

class PublicAddress extends Component {
  constructor(props) {
    super(props);
    this.inputRef = React.createRef();
    this.buttonRef = React.createRef();
  }

  handleClick = e => {
    const { inputRef, buttonRef } = this;

    e.preventDefault();
    inputRef.current.select();
    document.execCommand('copy');
    buttonRef.current.classList.add('-is-copied');

    setTimeout(() => {
      buttonRef.current.classList.remove('-is-copied');
    }, 5000);
  };

  render() {
    return (
      <Field label="Public address">
        <StyledWrapper>
          <StyledInput
            readOnly
            value={this.props.address}
            ref={this.inputRef}
          />
          <StyledButton
            size="small"
            px={3}
            onClick={this.handleClick}
            ref={this.buttonRef}
          >
            Copy
          </StyledButton>
        </StyledWrapper>
      </Field>
    );
  }
}

PublicAddress.displayName = 'PublicAddress';

export default PublicAddress;
