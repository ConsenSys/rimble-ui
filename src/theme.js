import { tint, shade } from 'polished';

const success = '#28C081';
const danger = '#DC2C10';
const warning = '#FD9D28';
const info = '#36ADF1';

// theme.js
export default {
  fontSizes: [12, 14, 16, 20, 24, 32, 48, 64],
  fontWeights: [0, 300, 400, 600, 700],
  letterSpacings: [0, 1, 2, 4, 8],
  lineHeights: {
    solid: 1,
    title: 1.25,
    copy: 1.5,
  },
  fonts: {
    serif: 'athelas, georgia, times, serif',
    sansSerif: '"Source Sans Pro", -apple-system, sans-serif',
  },
  space: [0, 4, 8, 16, 32, 64, 128, 256],
  radii: ['0', '4px', '8px', '16px'],
  width: [0, 16, 32, 64, 128, 256],
  minWidths: [0, 16, 32, 64, 128, 256],
  maxWidths: [0, 16, 32, 64, 128, 256, 512, 768, 1024, 1536],
  heights: [0, 16, 32, 64, 128, 256],
  minHeights: [0, 16, 32, 64, 128, 256],
  maxHeights: [0, 16, 32, 64, 128, 256],
  borders: [0, '1px solid transparent'],
  borderWidths: ['0', '1px', '2px', '4px'],
  shadows: [
    '0',
    '0px 2px 4px rgba(0, 0, 0, 0.1)',
    '0 7px 14px rgba(50,50,93,.1)',
  ],
  opacity: {
    disabled: 0.4,
  },
  colors: {
    primary: '#4E3FCE',
    blue: '#007ce0',
    navy: '#004175',
    copyColor: '#3F3D4B',
    // black: '#000',
    black: '#000e1a',
    'near-black': '#111',
    'dark-gray': '#333',
    'mid-gray': '#555',
    // gray: ' #777',
    grey: '#CCC',
    silver: '#999',
    'light-silver': '#aaa',
    'moon-gray': '#ccc',
    'light-gray': '#eee',
    'near-white': '#f4f4f4',
    white: '#fff',
    transparent: 'transparent',
    blacks: [
      'rgba(0,0,0,.0125)',
      'rgba(0,0,0,.025)',
      'rgba(0,0,0,.05)',
      'rgba(0,0,0,.1)',
      'rgba(0,0,0,.2)',
      'rgba(0,0,0,.3)',
      'rgba(0,0,0,.4)',
      'rgba(0,0,0,.5)',
      'rgba(0,0,0,.6)',
      'rgba(0,0,0,.7)',
      'rgba(0,0,0,.8)',
      'rgba(0,0,0,.9)',
    ],
    whites: [
      'rgba(255,255,255,.0125)',
      'rgba(255,255,255,.025)',
      'rgba(255,255,255,.05)',
      'rgba(255,255,255,.1)',
      'rgba(255,255,255,.2)',
      'rgba(255,255,255,.3)',
      'rgba(255,255,255,.4)',
      'rgba(255,255,255,.5)',
      'rgba(255,255,255,.6)',
      'rgba(255,255,255,.7)',
      'rgba(255,255,255,.8)',
      'rgba(255,255,255,.9)',
    ],
    danger: danger,
    warning: warning,
    success: success,
    info: info,
  },
  zIndices: [0, 9, 99, 999, 9999],
  messageStyle: {
    base: {
      color: shade(0.4, '#AAA'),
      backgroundColor: tint(0.9, '#AAA'),
      borderColor: '#AAA',
    },
    success: {
      color: shade(0.4, success),
      backgroundColor: tint(0.9, success),
      borderColor: success,
    },
    warning: {
      color: shade(0.4, warning),
      backgroundColor: tint(0.9, warning),
      borderColor: warning,
    },
    danger: {
      color: shade(0.4, danger),
      backgroundColor: tint(0.9, danger),
      borderColor: danger,
    },
    info: {
      color: shade(0.4, info),
      backgroundColor: tint(0.9, info),
      borderColor: info,
    },
  },
};
