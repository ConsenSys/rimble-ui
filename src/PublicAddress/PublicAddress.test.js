import PublicAddress from "./";
import ThemeProvider from "../ThemeProvider";
import theme from "../theme";

describe("PublicAddress component sanity", () => {
  it("has name", () => {
    expect(PublicAddress.displayName).toBe("PublicAddress");
  });

  it("matches default snapshot", () => {
    const component = render(<PublicAddress />);
    expect(component).toMatchSnapshot();
  });

  it("matches themed snapshot", () => {
    const component = render(
      <ThemeProvider theme={theme}>
        <PublicAddress />
      </ThemeProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
