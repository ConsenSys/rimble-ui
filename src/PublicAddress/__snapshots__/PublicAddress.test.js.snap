// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicAddress component sanity matches default snapshot 1`] = `
.c0 {
  margin-bottom: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  text-align: left;
}

.c0 {
  box-sizing: border-box;
}

.c5 {
  margin: 0px;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 0px;
  padding-bottom: 0px;
  color: #fff;
  background-color: #4E3FCE;
  font-size: inherit;
  height: 48px;
  height: small;
  width: small;
  position: relative;
  border: 0;
  border-color: none;
  border-radius: 4px;
  text-align: left;
}

.c5 {
  box-sizing: border-box;
}

.c1 {
  margin-bottom: 8px;
  color: #3F3D4B;
  font-size: 14px;
  text-align: left;
  font-size: 14px;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  font-weight: 600;
  text-align: left;
  line-height: 1.5;
}

.c1 {
  box-sizing: border-box;
}

.c1:after {
  content: ' (optional)';
  font-weight: normal;
  font-style: italic;
}

.c3 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 0px;
  padding-bottom: 0px;
  color: #3F3D4B;
  background-color: #fff;
  font-size: 1rem;
  height: 3rem;
  border: 1px solid transparent;
  border-color: #CCC;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 1rem;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  text-overflow: ellipsis;
  white-space: no-wrap;
}

.c3 {
  box-sizing: border-box;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c3:hover {
  box-shadow: 0px 2px 6px rgba(0,0,0,0.3);
}

.c3:focus {
  outline: none;
  border-color: #4E3FCE;
}

.c3:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.was-validated .c3:valid {
  border-color: #28C081;
}

.was-validated .c3:valid ~ svg.icon-valid {
  visibility: visible;
}

.was-validated .c3:invalid {
  border-color: #EC9081;
}

.was-validated .c3:invalid ~ svg.icon-invalid {
  visibility: visible;
}

.c4 {
  font-size: inherit;
  font-weight: 600;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 12px;
  height: 2rem;
  min-width: 2rem;
  padding: 0 1rem;
}

.c4 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  line-height: 1;
  overflow: hidden;
  white-space: nowrap;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all .15s ease;
  transition: all .15s ease;
  background-color: #4E3FCE;
}

.c4:hover {
  background-color: #5f52d2;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c4:active {
  background-color: #4638b9;
  box-shadow: none;
}

.c4:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c4 > svg {
  display: block;
}

.c4 > svg:first-child {
  margin-left: -1rem;
  margin-right: .5rem;
}

.c4 > svg:last-child {
  margin-left: .5rem;
  margin-right: -1rem;
}

.c4 > svg:first-child:last-child {
  margin: 0;
}

.c4 {
  position: relative;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4.-is-copied::before {
  content: '';
  position: absolute;
  width: 100%;
  height: 100%;
  background: inherit;
  opacity: 1;
  -webkit-animation-name: eMLfYp;
  animation-name: eMLfYp;
  -webkit-animation-duration: 100ms;
  animation-duration: 100ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

.c4::after {
  opacity: 0;
  content: 'Copied!';
  position: absolute;
  width: 100%;
  background: inherit;
  -webkit-transition: all 500ms ease;
  transition: all 500ms ease;
}

.c4.-is-copied::after {
  opacity: 1;
  -webkit-animation-name: kcJLrX;
  animation-name: kcJLrX;
  -webkit-animation-duration: 1000ms;
  animation-duration: 1000ms;
  -webkit-animation-timing-function: ease;
  animation-timing-function: ease;
}

.c2 {
  text-align: left;
}

.c2 {
  box-sizing: border-box;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  position: relative;
}

.c2 > input {
  width: 100%;
  padding-right: 5.25rem;
}

.c2 > button {
  position: absolute;
  right: 0.75rem;
}

<label
  class="c0"
  display="flex"
>
  <div
    class="c1"
    color="copyColor"
    font-family="sansSerif"
    font-size="1"
    font-weight="3"
  >
    Public Address
  </div>
  <div
    class="c2"
  >
    <input
      class="c3"
      color="copyColor"
      font-family="sansSerif"
      font-size="1rem"
      height="3rem"
      readonly=""
    >
      
    </input>
    <button
      class="c4 c5"
      color="white"
      font-family="sansSerif"
      font-size="inherit"
      font-weight="3"
      height="48px"
      type="button"
    >
      <span>
        Copy
      </span>
    </button>
  </div>
</label>
`;

exports[`PublicAddress component sanity matches themed snapshot 1`] = `
.c1 {
  margin-bottom: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  text-align: left;
}

.c1 {
  box-sizing: border-box;
}

.c6 {
  margin: 0px;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 0px;
  padding-bottom: 0px;
  color: #fff;
  background-color: #4E3FCE;
  font-size: inherit;
  height: 48px;
  height: small;
  width: small;
  position: relative;
  border: 0;
  border-color: none;
  border-radius: 4px;
  text-align: left;
}

.c6 {
  box-sizing: border-box;
}

.c2 {
  margin-bottom: 8px;
  color: #3F3D4B;
  font-size: 14px;
  text-align: left;
  font-size: 14px;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  font-weight: 600;
  text-align: left;
  line-height: 1.5;
}

.c2 {
  box-sizing: border-box;
}

.c2:after {
  content: ' (optional)';
  font-weight: normal;
  font-style: italic;
}

.c4 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 0px;
  padding-bottom: 0px;
  color: #3F3D4B;
  background-color: #fff;
  font-size: 1rem;
  height: 3rem;
  border: 1px solid transparent;
  border-color: #CCC;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 1rem;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  text-overflow: ellipsis;
  white-space: no-wrap;
}

.c4 {
  box-sizing: border-box;
}

.c4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c4:hover {
  box-shadow: 0px 2px 6px rgba(0,0,0,0.3);
}

.c4:focus {
  outline: none;
  border-color: #4E3FCE;
}

.c4:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.was-validated .c4:valid {
  border-color: #28C081;
}

.was-validated .c4:valid ~ svg.icon-valid {
  visibility: visible;
}

.was-validated .c4:invalid {
  border-color: #EC9081;
}

.was-validated .c4:invalid ~ svg.icon-invalid {
  visibility: visible;
}

.c5 {
  font-size: inherit;
  font-weight: 600;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 12px;
  height: 2rem;
  min-width: 2rem;
  padding: 0 1rem;
}

.c5 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  line-height: 1;
  overflow: hidden;
  white-space: nowrap;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all .15s ease;
  transition: all .15s ease;
  background-color: #4E3FCE;
}

.c5:hover {
  background-color: #5f52d2;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c5:active {
  background-color: #4638b9;
  box-shadow: none;
}

.c5:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c5 > svg {
  display: block;
}

.c5 > svg:first-child {
  margin-left: -1rem;
  margin-right: .5rem;
}

.c5 > svg:last-child {
  margin-left: .5rem;
  margin-right: -1rem;
}

.c5 > svg:first-child:last-child {
  margin: 0;
}

.c5 {
  position: relative;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5.-is-copied::before {
  content: '';
  position: absolute;
  width: 100%;
  height: 100%;
  background: inherit;
  opacity: 1;
  -webkit-animation-name: eMLfYp;
  animation-name: eMLfYp;
  -webkit-animation-duration: 100ms;
  animation-duration: 100ms;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

.c5::after {
  opacity: 0;
  content: 'Copied!';
  position: absolute;
  width: 100%;
  background: inherit;
  -webkit-transition: all 500ms ease;
  transition: all 500ms ease;
}

.c5.-is-copied::after {
  opacity: 1;
  -webkit-animation-name: kcJLrX;
  animation-name: kcJLrX;
  -webkit-animation-duration: 1000ms;
  animation-duration: 1000ms;
  -webkit-animation-timing-function: ease;
  animation-timing-function: ease;
}

.c3 {
  text-align: left;
}

.c3 {
  box-sizing: border-box;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  position: relative;
}

.c3 > input {
  width: 100%;
  padding-right: 5.25rem;
}

.c3 > button {
  position: absolute;
  right: 0.75rem;
}

.c0 {
  font-size: 1em;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  color: #3F3D4B;
}

<div
  class="c0"
>
  <label
    class="c1"
    display="flex"
  >
    <div
      class="c2"
      color="copyColor"
      font-family="sansSerif"
      font-size="1"
      font-weight="3"
    >
      Public Address
    </div>
    <div
      class="c3"
    >
      <input
        class="c4"
        color="copyColor"
        font-family="sansSerif"
        font-size="1rem"
        height="3rem"
        readonly=""
      >
        
      </input>
      <button
        class="c5 c6"
        color="white"
        font-family="sansSerif"
        font-size="inherit"
        font-weight="3"
        height="48px"
        type="button"
      >
        <span>
          Copy
        </span>
      </button>
    </div>
  </label>
</div>
`;
